### PR TITLE
End game mechanics

### LINKED ISSUE #978: improve UI for win/loss
Currently just a message in the info panel. Ought to have a pop up and disable the next turn/takeoff buttons.

### PR DESCRIPTION
**Pull request by Raffson**
Fixes #978

### CODE REVIEW

**Comment by DanAlbert** [comment]
Sorry for the delays in getting your PRs reviewed. A handful of us are fairly busy and haven't had as much time to spend here lately. We'll get to them, just wanted to let you know that they aren't being ignored.

**Comment by Raffson** [comment]
No worries, perhaps in time I might be able to help 😉

I'll see to it that the lint-error gets resolved in the next 24 hours.

**Comment by Raffson** [comment]
I believe the lint-errors are resolved, waiting for approval of workflows.

Also, it might be handy if I could add you on discord so I can discuss or ask stuff via PM. Please keep an eye out for hashtag 2466.

**Comment by DanAlbert** [comment]
Everyone going through my DMs scales terribly, but I think I found you and added you to the dev channel :)

**Comment by Raffson** [comment]
After I ran `black` it turned out I forgot a couple of commas, so this PR should be done. Although as pointed out in the linked issue, with the transition to a "turnless" campaign the `SimSpeedControls` should be reviewed upon "end-game", but it's probably best if we leave that for later. Perhaps with its own issue and PR.

**Comment by Raffson** [comment]
Converted back to draft as I noticed a subtle bug that needs fixing:

When capturing the last point with the cheat-function the behavior is as expected, however the second time (basically undoing your last action) all packages suddenly disappear from the list, though the flight plans are still visible on the map. IMO this shouldn't happen unless you actually flew the mission, but then the flight plans should also be removed from the map...

**Comment by Raffson** [comment]
Ok just completed a "live" test, the popup shows as expected. The only problem left is the disappearing of packages when "cheating your way back into the game". Now there's 2 possible fixes that seem viable for this and I'm wondering which one is preferred.

Option 1: prevent cheats when a campaign is done, this way the user can't "revert back"...
Option 2: fix the actual issue so that flights don't disappear from the package list when "reverting back" after the campaign ended.

While the first option will probably be easier, it's more of a constraining solution, though one might ask why would you want to use option 2 at all. If anything, the user should save a backup right before winning the campaign if he wishes "revert back".

Any thoughts?

Edit: also concluding that the signal for ending the campaign would sit better in de debriefing window instead of the "waiting for results" window.

File: game/game.py
Comment (Raffson) [inline game/game.py:390]: If we don't check for the win/loss condition first, we'll get an exception in self.set_bullseye() which should be fairly obvious since bullseye is set on the closest enemy CP resepectively and in case of a win/loss one of the the teams has no CPs left. It also speaks for itself that there's no need to set the bullseyes anymore after a campaign is done...

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py:82]: Need to reference this in `setGame` to lock the controls...

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py:134]: `if game` <=> `if game is not None` for which we have already tested at the beginning of the method, no need to check it again...

File: qt_ui/widgets/simspeedcontrols.py
Comment (Raffson) [inline qt_ui/widgets/simspeedcontrols.py:36]: Override `setEnabled` to enable/disable the buttons for controlling the sim speed.

File: qt_ui/windows/QDebriefingWindow.py
Comment (Raffson) [inline qt_ui/windows/QDebriefingWindow.py:91]: Upon closing of the debriefing window, check for win/loss and emit signal if required.

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [inline qt_ui/windows/QLiberationWindow.py:374]: End-game message, closing all windows except main, start new campaign wizard if user chooses to start a new campaign.

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py:133]: Locking controls in QTopPanel, this is one of the things I'm considering to refactor to make the code a little cleaner...

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [inline qt_ui/windows/QLiberationWindow.py:395]: This is needed to let the map fully update <ins>after the last turn</ins> since the next line will "release" the game. If that happens too soon, the map doesn't update properly and will still depict the image from the last turn, i.e. <ins>before the last turn</ins>. It's specifically the case with a cheat-capture, less applicable to winning/losing through the debriefing window, i.e. winning/losing after actually flying the "last mission".

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [inline qt_ui/windows/QLiberationWindow.py:396]: This prevents the user from clicking on the map, opening CP's and what not, to possibly revert the state of the game through the cheat button. This also prevents the user from buying units or transferring them, which would be useless if the campaign has ended. It also removes all packages and flights from the ATO panel which is what we want, or at least that's what I assume.
One small issue I noticed is that it's possible one of the flight-plans might get selected on the map, allowing you to move the markers around even though the flight doesn't exist anymore. The lines of the flight-plan stay drawn the same way, it's literally only the pin-marker that becomes movable. Now as far as I know, this shouldn't cause any exceptions so I think leaving it as it is is a trade-off we should consider.

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [inline qt_ui/windows/QLiberationWindow.py:554]: Force debriefing to go out of scope, otherwise the loop over `QApplication.topLevelWidgets()` will close the debriefing window again which may trigger the end-game message for a second time. Mind that it shouldn't be possible to close the main window when a debriefing window is still open because the debriefing window forces the focus on itself for as long as it is open. This means we can let the debriefing go out of scope safely.

File: game/game.py
Comment (Raffson) [inline game/game.py:390]: If we don't check for the win/loss condition first, we'll get an exception in self.set_bullseye() which should be fairly obvious since bullseye is set on the closest enemy CP resepectively and in case of a win/loss one of the the teams has no CPs left. It also speaks for itself that there's no need to set the bullseyes anymore after a campaign is done...

File: game/game.py
Comment (Raffson) [inline game/game.py:394]: If we don't check for the win/loss condition first, we'll get an exception in self.set_bullseye() which should be fairly obvious since bullseye is set on the closest enemy CP resepectively and in case of a win/loss one of the the teams has no CPs left. It also speaks for itself that there's no need to set the bullseyes anymore after a campaign is done...

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py:86]: Need to reference this in `setGame` to lock the controls...

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py:139]: Locking controls in `QTopPanel`, this is one of the things I'm considering to refactor to make the code a little cleaner...

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py:148]: `if game` <=> `if game is not None` for which we have already tested at the beginning of the method, no need to check it again...

File: qt_ui/widgets/simspeedcontrols.py
Comment (Raffson) [inline qt_ui/widgets/simspeedcontrols.py:39]: Override `setEnabled` to enable/disable the buttons for controlling the sim speed.

File: qt_ui/windows/GameUpdateSignal.py
Comment (Raffson) [inline qt_ui/windows/GameUpdateSignal.py:42]: End-game signal emitter...

File: qt_ui/windows/QDebriefingWindow.py
Comment (Raffson) [inline qt_ui/windows/QDebriefingWindow.py:96]: Upon closing of the debriefing window, check for win/loss and emit signal if required.

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [inline qt_ui/windows/QLiberationWindow.py]: End-game message, closing all windows except main, update game with `None` to prevent the user for manipulating the campaign further, start new campaign wizard if user chooses to do so.

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [inline qt_ui/windows/QLiberationWindow.py]: `sleep(0.5)` is needed to let the map fully update <ins>after the last turn</ins> since the next line will "release" the game. If that happens too soon, the map doesn't update properly and will still depict the image from the last turn, i.e. <ins>before the last turn</ins>. It's specifically the case with a cheat-capture, less applicable to winning/losing through the debriefing window, i.e. winning/losing after actually flying the "last mission".

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [inline qt_ui/windows/QLiberationWindow.py:382]: `GameUpdateSignal.get_instance().updateGame(None)` prevents the user from clicking on the map, opening CP's and what not, to possibly revert the state of the game through the cheat button. This also prevents the user from buying units or transferring them, which would be useless if the campaign has ended. It also removes all packages and flights from the ATO panel which is what we want, or at least that's what I assume.
One small issue I noticed is that it's possible one of the flight-plans might get selected on the map, allowing you to move the markers around even though the flight doesn't exist anymore. The lines of the flight-plan stay drawn the same way, it's literally only the pin-marker that becomes movable. Now as far as I know, this shouldn't cause any exceptions so I think leaving it as it is is a trade-off we should consider.

File: qt_ui/windows/basemenu/QBaseMenu2.py
Comment (Raffson) [inline qt_ui/windows/basemenu/QBaseMenu2.py:132]: End-game signaling logic for the "cheat-capture" button

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py:99]: Bundle all controls that should be locked when a campaign ends

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py:144]: Lock/Unlock controls

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py:134]: `if game` <=> `if game is not None` for which we have already tested at the beginning of the method, no need to check it again...

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py:151]: When the campaign hasn't started yet, lock time controls

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [inline qt_ui/widgets/QTopPanel.py]: Raise an `AssertionError ` if `game.turn` < 0, because this should really NEVER happen, so if it does, we'd like to know...

File: qt_ui/widgets/QTopPanel.py
Comment (DanAlbert) [inline qt_ui/widgets/QTopPanel.py:92]: I think this would be better inverted. That is, add a signal for game state (which might just be the playing/ended boolean, but maybe useful to have the full set of not loaded/turn 0/playing/over?) and move the logic for enable/disable into the components themselves rather than the owner.

WDYT?

File: qt_ui/widgets/QTopPanel.py
Comment (DanAlbert) [inline qt_ui/widgets/QTopPanel.py]: Needs a comment explaining why this would happen. Are you trying to re-disable things if someone loads an already completed game? Or is this one of those `setGame` calls that actually happens every turn?

OTOH, if you take my suggestion to invert the ownership, this confusing bit probably just disappears.

File: qt_ui/widgets/QTopPanel.py
Comment (DanAlbert) [reply qt_ui/widgets/QTopPanel.py]: Yes we would, so raise `RuntimeError` rather than using `assert`, which is disabled in release builds. I think the only places `assert` is useful is for forcing the type system to acknowledge an otherwise inexpressible constraint, and tests.

File: qt_ui/widgets/simspeedcontrols.py
Comment (DanAlbert) [inline qt_ui/widgets/simspeedcontrols.py:37]: If you override this you need to call the base method, and you also need to override `setDisabled`.

File: qt_ui/windows/GameUpdateSignal.py
Comment (DanAlbert) [inline qt_ui/windows/GameUpdateSignal.py]: Oh, you _do_ have this signal. Probably better to respond to it in each component that needs to rather than in their parent. We don't want to accidentally lose behavior just because a parent changed.

Comment (DanAlbert) [inline qt_ui/windows/GameUpdateSignal.py]:
```suggestion
    game_state_changed = Signal(TurnState)
```

File: qt_ui/windows/QDebriefingWindow.py
Comment (DanAlbert) [inline qt_ui/windows/QDebriefingWindow.py:94]: Why is this needed? Isn't this going to happen in `initialize_turn`? I'd rather the UI be responsible for as little game state management as possible.

File: qt_ui/windows/QLiberationWindow.py
Comment (DanAlbert) [reply qt_ui/windows/QLiberationWindow.py]: If that's a problem you need real synchronization. I'm not convinced we should be clearing the map though. Players probably want to see what they've accomplished.

File: qt_ui/windows/QLiberationWindow.py
Comment (DanAlbert) [reply qt_ui/windows/QLiberationWindow.py:554]: You can't assume anything about when this will be garbage collected.

File: qt_ui/windows/basemenu/QBaseMenu2.py
Comment (DanAlbert) [inline qt_ui/windows/basemenu/QBaseMenu2.py:130]: Why is this being duplicated? Shouldn't this all be handled in a single handler that responds to `updateGame`?

**Review by DanAlbert — CHANGES_REQUESTED** [review]

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [reply qt_ui/windows/QLiberationWindow.py]: This is probably obsolete since the PBO, because now we should be pushing all the last changes before releasing the game, but I'd have to double check this...

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [reply qt_ui/windows/QLiberationWindow.py:554]: > You can't assume anything about when this will be garbage collected.

That's not the point... This should assure that the end-game popup doesn't show itself a second time, because it will if you leave this line out.

Either way, once I reviewed everything and all changes are applied, I'll verify if this is still an issue.

File: qt_ui/windows/basemenu/QBaseMenu2.py
Comment (Raffson) [reply qt_ui/windows/basemenu/QBaseMenu2.py:130]: I'll have a look at it, initially the reasoning was the following:

The game can end in 2 scenarios, the first being a cheat-capture of the last enemy/friendly base. The second being a normal capture event of the last base, thus upon debriefing.

That said, it's possible that I missed something that these 2 scenarios have in common, and intuitively `updateGame` seems like a winner so I'll check it out...

File: qt_ui/widgets/simspeedcontrols.py
Comment (Raffson) [reply qt_ui/widgets/simspeedcontrols.py:37]: Calling the base method, yes, ~~but now that I'm testing this again it seems like it's not working, with or without the 'super-call'...~~
EDIT: nvm, I'm just starting to forget things -_-

Override setDisabled, no, because it doesn't exist. (simSpeedControls is derived from QHBoxLayout)

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [reply qt_ui/widgets/QTopPanel.py:92]: I think that would make things less readable. If I understood what you're saying, it would require connecting the `game_state_changed` signal to all of those controls separately. Let alone the fact that we need to check whether or not the campaign has started, because that affects the state of some of the controls as well.

Either way, I'll look into the whole refactor thing together with the checks for the game state that are responsible for emitting the signals.

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [reply qt_ui/widgets/QTopPanel.py]: Done

File: qt_ui/windows/GameUpdateSignal.py
Comment (Raffson) [reply qt_ui/windows/GameUpdateSignal.py]: Manually changed...

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [reply qt_ui/widgets/QTopPanel.py]: > Are you trying to re-disable things if someone loads an already completed game? Or is this one of those `setGame` calls that actually happens every turn?

By releasing the game in `QLiberationWindow.onEndGame`, the user shouldn't be able to save the game anymore. Hence, loading a game that's completed should be impossible.

`setGame` is connected to the `gameupdated` signal, and I assume we'd want to check if the game ended after each update. That brings us back to integrating it into the update event, which is what I'm checking next.

File: qt_ui/windows/QDebriefingWindow.py
Comment (Raffson) [reply qt_ui/windows/QDebriefingWindow.py:94]: Deleted, refactoring to `game.process_win_loss`

File: qt_ui/windows/basemenu/QBaseMenu2.py
Comment (Raffson) [reply qt_ui/windows/basemenu/QBaseMenu2.py:130]: Refactored to `game.process_win_loss`

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [reply qt_ui/windows/QLiberationWindow.py]: Confirmed, sleep is no longer needed. Makes sense if you think about it because by the time the front-end started querying the back-end (as it used to be), the back-end already released the game and so nothing updated after winning/losing.

Comment (Raffson) [reply qt_ui/windows/QLiberationWindow.py]: > If that's a problem you need real synchronization. I'm not convinced we should be clearing the map though. Players probably want to see what they've accomplished.

The map doesn't clear by "releasing the game", but it does disable all TGO so you can't open dialogs by clicking on them anymore...

File: qt_ui/windows/GameUpdateSignal.py
Comment (Raffson) [reply qt_ui/windows/GameUpdateSignal.py]: Not sure what you mean...

File: qt_ui/windows/QDebriefingWindow.py
Comment (Raffson) [reply qt_ui/windows/QDebriefingWindow.py:94]: Ok now I remember why I did this. The check in "initialize turn" happens sooner, so if I emit the signal from there we get the win/loss message a little too soon imo. I even doubt if we'll get to see the debriefing window because I force all windows to be closed except the main window.
I'll see if I can find a way around this...

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [reply qt_ui/widgets/QTopPanel.py:92]: I still think it's best to handle stuff here, however the logic in `setGame` is better off in a separate method ~~that's linked to the `game_state_changed` signal~~, which I already did on my end 👍 

EDIT: linking the signal doesn't help out much because `setGame` will be called again after the end-game signal was triggered, which will re-enable certain buttons that we don't want...

File: qt_ui/widgets/QTopPanel.py
Comment (Raffson) [reply qt_ui/widgets/QTopPanel.py]: Refactored this into a method called `setControls`, connecting it with the `game_state_changed` signal.

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [reply qt_ui/windows/QLiberationWindow.py:554]: Conclusion is the following: removing the `closeEvent` method in `QDebriefingWindow` fixes the "double pop-up" issue, however as mentioned in the related conversation this does introduce its own set of problems.

File: qt_ui/windows/basemenu/QBaseMenu2.py
Comment (Raffson) [reply qt_ui/windows/basemenu/QBaseMenu2.py:130]: I've been at it for some time now but I keep hitting a brick wall. What I had assured the signal would only be triggered once. With your suggestions I can't seem to guarantee that...

File: qt_ui/windows/QDebriefingWindow.py
Comment (Raffson) [reply qt_ui/windows/QDebriefingWindow.py:94]: Turns out the debriefing window does actually show, but I can't seem to find a better spot to trigger the signal without guaranteeing it's only triggered once...

File: qt_ui/windows/basemenu/QBaseMenu2.py
Comment (Raffson) [reply qt_ui/windows/basemenu/QBaseMenu2.py:130]: Changing this:
```
GameUpdateSignal.get_instance().updateGame(self.game_model.game)
state = self.game_model.game.check_win_loss()
if state == TurnState.WIN or state == TurnState.LOSS:
    GameUpdateSignal.get_instance().endGame(state)
```
into:
```
self.close()
```

Would work for cheat-capture given we refactor the signal-trigger to `QLiberationWindow.setGame`, because upon closing an update is signaled. Also, closing seems like a reasonable approach after a cheat-capture as the UI state is invalid for the CP-window. ~~As for normal capture events I'm still having a double pop-up...~~
As for normal capture events, it seems like "releasing the game" happens twice, which wipes the map. If I can figure out a way to prevent that and "release the game" like it happens with a cheat-capture, then I think I may have found a solution after all...

Comment (Raffson) [reply qt_ui/windows/basemenu/QBaseMenu2.py:130]: Well after closer inspection it's not the game that gets released twice, but rather an update that gets sent to the front-end, which is exactly what I don't want as this wipes the map. I'm giving it one last try to refactor these signal triggers but if that doesn't work out, I'm recommending leaving the signal triggers where they are...

File: qt_ui/windows/QDebriefingWindow.py
Comment (Raffson) [reply qt_ui/windows/QDebriefingWindow.py:94]: > Why is this needed?

Two cases a game can end, after a cheat-capture and after a debriefing, this is the latter...

> Isn't this going to happen in `initialize_turn`?

Yes, but that would mean calling `GameUpdateSignal.get_instance().gameStateChanged`. I'm not a fan of that because that means UI stuff is making its way into the model. The main reason the check needs to happen in `initialize_turn` is to actually prevent turn initialization, because in an end-game situation that'll cause a bunch of exceptions.


> I'd rather the UI be responsible for as little game state management as possible.

Please elaborate because I really don't see how I'm "managing game state" by calling `check_win_loss`

File: qt_ui/windows/QLiberationWindow.py
Comment (Raffson) [reply qt_ui/windows/QLiberationWindow.py:554]:
```suggestion
            if hasattr(self, 'debriefing'):
                del self.debriefing
```

If this makes you happier, it's an alternative I've tested which also works...

File: qt_ui/windows/GameUpdateSignal.py
Comment (Raffson) [reply qt_ui/windows/GameUpdateSignal.py]: > Oh, you _do_ have this signal.

Euhm, *duh*...

> Probably better to respond to it in each component that needs to rather than in their parent.

That'll require more work, since you mentioned on discord that you don't even think this is worth solving until qt is gone, then I'm not sure what the problem is...

> We don't want to accidentally lose behavior just because a parent changed.

What behavior? If this signal is triggered, there shouldn't be any behavior since the campaign ended. We only need to assure the user can't do anything that might possibly revert back the state. From my understanding that means doing the following:

- Release the game by triggering the `gameupdated` signal with `None`, this has the following effects:
  * Disables all "game-actions" (save game, options, stats and notes, i.e. icons in the top-left corner)
  * "Disables" the map, i.e. none of the TGOs/CPs can be opened anymore
  * Wipes the ATO panel clean
- Disable buttons in QTopPanel:
  * Air Wing, as this might create ferry flights
  * Transfers, a precaution because we released the game
  * SimSpeedControls, also a precaution
  * Pass Turn, finishing/initializing a turn when the game is "released" is definitely a problem
  * Take Off, same as Pass Turn but then under the context of mission generation
 
The combination of both points ought to result in a "frozen" environment, which I assume is what we want when a campaign ends...

**Pull request closed without merging**
